theano/tensor/opt.py

```diff
@@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
 def local_grad_clip(node):
     if isinstance(node.op, theano.gradient.GradClip):
         return node.inputs
+
+@register_canonicalize
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.Alloc])
+def local_reduce_double_alloc(node):
+    # Alloc(Alloc(m, x, 1), x, y) -> Alloc(m, x, y)
+    if not isinstance(node.op, T.Alloc):
+        return False
+    if not node.inputs[0].owner or not isinstance(node.inputs[0].owner.op, T.Alloc):
+        return False
+    inputs_outer = node.inputs
+    inputs_inner = node.inputs[0].owner.inputs
+    if len(inputs_outer) != len(inputs_inner):
+        return False
```

Comment:
Do you mean T.alloc(T.alloc(m, x, y, 1, 1), x, y, z, w) instead of
T.alloc(T.alloc(m, y, 1, 1), x, y, z, w)? There is no x in inner alloc, so
how does it get broadcasted in the outer alloc?


On Mon, Jun 29, 2015 at 4:26 PM, Frédéric Bastien <notifications@github.com>
wrote:

> In theano/tensor/opt.py
> <https://github.com/Theano/Theano/pull/3082#discussion_r33508570>:
>
> > @@ -5841,3 +5841,26 @@ def local_add_mul_fusion(node):
> >  def local_grad_clip(node):
> >      if isinstance(node.op, theano.gradient.GradClip):
> >          return node.inputs
> > +
> > +@register_canonicalize
> > +@register_stabilize
> > +@register_specialize
> > +@gof.local_optimizer([T.Alloc])
> > +def local_reduce_double_alloc(node):
> > +    # Alloc(Alloc(m, x, 1), x, y) -> Alloc(m, x, y)
> > +    if not isinstance(node.op, T.Alloc):
> > +        return False
> > +    if not node.inputs[0].owner or not isinstance(node.inputs[0].owner.op, T.Alloc):
> > +        return False
> > +    inputs_outer = node.inputs
> > +    inputs_inner = node.inputs[0].owner.inputs
> > +    if len(inputs_outer) != len(inputs_inner):
> > +        return False
>
> optimized
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3082/files#r33508570>.
>